### PR TITLE
Update findings endpoint to include outdated component findings

### LINKED ISSUE #2003: Project level recommendation to "Safer & Closest Version" or "Safest Version" to fix an existing vulnerable component version
The enhancement may already be reported! Please search for the enhancement before creating one.

### Current Behavior:
Specific to SCA, Dependency-Track already supports outdated version detection. However, at the project level, when one visits the "Audit Vulnerabilities" tab, insights of the components that are vulnerable with the current version, severity, GHSA / NVD insights etc. are available but this diagnosis has limited value without any recommendation to retrofit the component.  that is commonly available in most SCA tools and platforms 

### Proposed Behavior:
It would be nice to have Dependency Track recommend or provide insights to the "Safer & Closest Version" or "Safest Version" for the vulnerable component version in use. This will help users with an actionable outcome as part of remediating the vulnerability rather than looking at other SCA tools and external sources on what version to upgrade / downgrade to fix the vulnerable library.

### LINKED ISSUE #2514: Report outdated component findings
### Current Behavior:

Outdated Components is in the OWASP Top 10: A06:2021 – Vulnerable and Outdated Components

Dependency Track knows of outdated components, but does not report them as findings

### Proposed Behavior

Create an endpoint to report the outdated component findings. Outdated component findings are Components that are direct dependencies of the project and have a higher stable version (see #2500)  available. Indirect dependencies should not be reported as one would normally not update these. Indirect dependencies are reported by the vulnerabilities findings

For each component the highest stable version is reported together with the last check date and optionally the published date of the new version.

Partly fixes #2003, #1833 

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

### PR DESCRIPTION
**Pull request by walterdeboer**
### Description

Update findings endpoint to include outdated component findings
* **/findings/project/{uuid}** both outdated components and components with vulnerabilities

Two new endpoints: 
* **/findings/project/{uuid}/vulnerabilities** : only components with vulnerabilities
* **/findings/project/{uuid}/outdated**, only outdated components (with vulnerabilities when found)

N.B. only direct dependencies are reported as these are the ones you could update yourself

### Addressed Issue

closes #2514, partly fixes #2003, #1833

### Additional Details

I could have aded some more params to the findings endpoint and leave the original output unchanged (backwards compatible for users)? No need for extra endpoints then. Because getFindins now includes outdated components, this also affects FindingsUploaders

This PR is based on (and needs) https://github.com/DependencyTrack/dependency-track/pull/2501 although it could be an independend enhancement

No documentation updated, I think exisiting documentation already covers this.

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [ ] This PR fixes a defect, and I have provided tests to verify that the fix is effective
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [ ] This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)
- [x] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly

### CODE REVIEW

File: src/main/java/org/dependencytrack/persistence/FindingsQueryManager.java [hunk 1]
```diff
@@ -264,36 +307,102 @@ public List<Finding> getFindings(Project project) {
...
+     */
+    @Override
+    public List<Finding> getOutdatedComponentFindings(Project project, boolean includeSuppressed, boolean withoutVulnerabilitiesOnly) {
+        final Query<Object[]> query = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, withoutVulnerabilitiesOnly ? Finding.QUERY_OUTDATED_WITHOUT_VULNERABILITIES : Finding.QUERY_OUTDATED);
+        query.setParameters(project.getId());
+        final List<Object[]> list = query.executeList();
+        final List<Finding> outdatedComponentFindings = new ArrayList<>();
+        for (final Object[] o: list) {
+            final Finding outdatedComponentFinding = new Finding(project.getUuid(), o);
+            final Component component = getObjectByUuid(Component.class, (String)outdatedComponentFinding.getComponent().get("uuid"));
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/persistence/FindingsQueryManager.java:358]: @sonatype-lift ignore

File: src/main/java/org/dependencytrack/persistence/FindingsQueryManager.java [hunk 2]
```diff
@@ -264,36 +307,102 @@ public List<Finding> getFindings(Project project) {
...
         query.setParameters(project.getId());
         final List<Object[]> list = query.executeList();
-        final List<Finding> findings = new ArrayList<>();
+        final List<Finding> vulnerabilityfindings = new ArrayList<>();
         for (final Object[] o: list) {
-            final Finding finding = new Finding(project.getUuid(), o);
-            final Component component = getObjectByUuid(Component.class, (String)finding.getComponent().get("uuid"));
-            final Vulnerability vulnerability = getObjectByUuid(Vulnerability.class, (String)finding.getVulnerability().get("uuid"));
+            final Finding vulnerabilityFinding = new Finding(project.getUuid(), o);
+            final Component component = getObjectByUuid(Component.class, (String)vulnerabilityFinding.getComponent().get("uuid"));
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/persistence/FindingsQueryManager.java:318]: @sonatype-lift ignore

File: src/main/java/org/dependencytrack/resources/v1/FindingResource.java [hunk 3]
```diff
@@ -172,5 +173,73 @@ public Response analyzeProject(
...
+                if (qm.hasAccess(super.getPrincipal(), project)) {
+                    final List<Finding> outdatedComponentFindingForProject = qm.getOutdatedComponentFindings(project, suppressed);
+                    return Response.ok(outdatedComponentFindingForProject).header(TOTAL_COUNT_HEADER, outdatedComponentFindingForProject.size()).build();
+                } else {
+                    return Response.status(Response.Status.FORBIDDEN).entity("Access to the specified project is forbidden").build();
+                }
+            } else {
+                return Response.status(Response.Status.NOT_FOUND).entity("The project could not be found.").build();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/resources/v1/FindingResource.java:207]: @sonatype-lift ignore

File: src/main/java/org/dependencytrack/tasks/repositories/MavenMetaAnalyzer.java [hunk 4]
```diff
@@ -110,4 +99,70 @@ public MetaModel analyze(final Component component) {
...
+     * The "latest" field is only intended for plugin resolution, and is only set upon deployment of a maven-plugin artifact, both for Maven 2.x and 3.x regardless whether a release or snapshot gets deployed.
+     *
+     * Also, Maven will update these fields with whatever version it is currently deploying, so "latest" and "release" will not necessarily correspond to the highest version number.
+     *
+     * https://support.sonatype.com/hc/en-us/articles/213464638-Why-are-the-latest-and-release-tags-in-maven-metadata-xml-not-being-updated-after-deploying-artifacts-
+     */
+    private void analyzeContent(final MetaModel meta, InputStream in)
+            throws SAXException, IOException, ParserConfigurationException, XPathExpressionException {
+
+        final Document document = XmlUtil.buildSecureDocumentBuilder().parse(in);
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/tasks/repositories/MavenMetaAnalyzer.java:120]: @sonatype-lift ignore

File: src/main/java/org/dependencytrack/policy/PolicyEngine.java [hunk 5]
```diff
@@ -58,6 +57,7 @@ public PolicyEngine() {
         evaluators.add(new ComponentAgePolicyEvaluator());
         evaluators.add(new ComponentHashPolicyEvaluator());
         evaluators.add(new CwePolicyEvaluator());
+        evaluators.add(new OutdatedComponentPolicyEvaluator());
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/policy/PolicyEngine.java]: oeps, wrong branch

File: src/main/java/org/dependencytrack/model/VersionDistance.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    @Override
+    public String toString() {
+        return major + "." + minor + "." + patch;
+    }
+
+    /**
+     * Calculates the distance between two versions
+     * @param version1
+     * @param version2
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/model/VersionDistance.java]: oeps, wrong branch

File: src/main/java/org/dependencytrack/policy/PolicyEngine.java [hunk 7]
```diff
@@ -58,6 +57,7 @@ public PolicyEngine() {
         evaluators.add(new ComponentAgePolicyEvaluator());
         evaluators.add(new ComponentHashPolicyEvaluator());
         evaluators.add(new CwePolicyEvaluator());
+        evaluators.add(new VersionDistancePolicyEvaluator());
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/policy/PolicyEngine.java]: oeps, wrong branch

File: src/main/java/org/dependencytrack/util/ComponentVersion.java [hunk 8]
```diff
@@ -18,115 +18,222 @@
...
+        "(?:" + DFSG_PATTERN_STRING + ")?" + DEBIAN_PATTERN_STRING + // debian release + optional DFSG part
+        "|" + DFSG_PATTERN_STRING + // DFSG part only
+        ")" + LABEL_PATTERN_STRING + BUILD_METADATA_PATTERN_STRING + "$",
+        Pattern.CASE_INSENSITIVE
+        );
+
+    // Semver version format:
+    // https://semver.org/
+    // restricted number, label and build metadata parts to preven ReDOS attacks.
+    protected static final Pattern SEMVER_PATTERN = Pattern.compile("^(0|[1-9]\\d{0,32})\\.(0|[1-9]\\d{0,32})\\.(0|[1-9]\\d{0,32})(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)){0,8}))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+){0,8}))?$");
```
[the remark this answers is not in the document]
Comment (walterdeboer) [reply src/main/java/org/dependencytrack/util/ComponentVersion.java:88]: @nscuro any thoughts on how to fix this? Or can it safely be ignored?

**Comment by stevespringett** [comment]
Findings are specific to vulnerabilities and all findings can be audited. Mixing up out-of-date components into findings is not recommended as out-of-date components cannot and should not be audited. These are two entirely different concepts.

**Pull request closed without merging**